core/contract.go

```diff
@@ -0,0 +1,49 @@
+package core
+
+import (
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+// Class unambiguously defines a [Contract]'s semantics.
+type Class struct {
+	// The version of the class, currently always 0.
+	APIVersion uint
+	// External functions defined in the class.
+	Externals []EntryPoint
+	// Functions that receive L1 messages. See
+	// https://www.cairo-lang.org/docs/hello_starknet/l1l2.html#receiving-a-message-from-l1
+	L1Handlers []EntryPoint
+	// Constructors for the class. Currently, only one is allowed.
+	Constructors []EntryPoint
+	// An ascii-encoded array of builtin names imported by the class.
+	Builtins []string
+	// The starknet_keccak hash of the JSON compiler output.
+	ProgramHash []byte
+	Bytecode    []*felt.Felt
+}
+
+// Hash computes the [Pedersen Hash] of the class.
+//
+// [Pedersen Hash]: https://docs.starknet.io/documentation/develop/Contracts/contract-hash/#how_the_class_hash_is_computed
+func (c *Class) Hash() {
+	// TODO
+}
+
+// Contract is an instance of a [Class].
+type Contract struct {
+	// The number of transactions sent from this contract.
+	// Only account contracts can have a non-zero nonce.
+	Nonce uint
+	// Hash of the class that this contract instantiates.
+	ClassHash *felt.Felt
+	// Root of the contract's storage trie.
+	StorageRoot *felt.Felt // TODO: is this field necessary?
```

Comment:
Added a comment about this [here](https://github.com/NethermindEth/juno/pull/426).

But yes, I believe this is necessary for smart contracts. The contract storage contains the smart contract data that should be accessible to users.

Reply:
```suggestion
	StorageRoot *felt.Felt 
```

Reply:
I agree that the storage should be stored somewhere, but I don't yet see a point with storing the storage root in the `Contract` struct. I can't think of a place it would be used.

Right now I am thinking we will only need to store the class hash and nonce with the contract instance. The contract storage will be stored separately (for performance). Both the hash+nonce and storage will be keyed on the contract address. But none of this is set in stone so feel free to suggest enhancements.